File: cracking_the_coding_interview/1_2_is_permutation/is_permutation.cpp
#include <unordered_map>
#include <string>
#include <cassert>
#include <iostream>

// time complexity: O(N)
// space complexity: O(N)
bool is_permutation(const std::string& input1, const std::string& input2) {
    if (input1.size() != input2.size()){
        return false;
    }

    std::unordered_map<char, int> counts;

    for(const auto c: input1) {
        counts[c]++;
    }

    for(const auto c: input2) {
        // will insert default constructed int when not existing, which is 0
        if(--counts[c] < 0) {
            return false;
        }
    }

    return true;
}


// test implementation

#define ASSERT(expression)                               \
    if(!(expression))                                    \
    {                                                    \
        std::cerr << "Assertion failed: " << #expression \
                  << "\nFile: " << __FILE__ << "\n"      \
                  << "\nLine: " << __LINE__ << "\n";     \
        return false;                                    \
    }


bool run_tests() {
    ASSERT(is_permutation("", ""));
    ASSERT(is_permutation("a", "a"));
    ASSERT(is_permutation("ab", "ab"));
    ASSERT(is_permutation("ab", "ba"));
    ASSERT(is_permutation("abcdefg12345", "gfedcba54321"));

    ASSERT(!is_permutation("", "1"));
    ASSERT(!is_permutation("a", "ab"));
    ASSERT(!is_permutation("a", "aaaaaa"));
    ASSERT(!is_permutation("ab", "abaaa"));
    ASSERT(!is_permutation("ab", "ab "));
    ASSERT(!is_permutation("abc", "abcd"));

    return true;
}


int main() {
    if(run_tests()) {
        std::cout << "C++ test succeeded\n";
        return 0;
    } else {
        std::cout << "Failure in C++ tests\n";
        return 1;
    }
}

File: cracking_the_coding_interview/1_2_is_permutation/is_permutation.rs
// time complexity: O(N)
// space complexity: O(N)
fn is_permutation(input1: &str, input2: &str) -> bool {
    if input1.len() != input2.len() {
        return false
    }

    let mut counts = std::collections::HashMap::new();

    for c in input1.chars() {
        let counter = counts.entry(c).or_insert(0);
        *counter += 1;
    }

    for c in input2.chars() {
        let counter = counts.entry(c).or_insert(0);
        *counter -= 1;

        if *counter < 0 {
            return false;
        }
    }

    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_permutation() {
        assert!(is_permutation("", ""));
        assert!(is_permutation("a", "a"));
        assert!(is_permutation("ab", "ab"));
        assert!(is_permutation("ab", "ba"));
        assert!(is_permutation("abcdefg12345", "gfedcba54321"));
    }

    #[test]
    fn test_is_not_permutation() {
        assert!(!is_permutation("", "1"));
        assert!(!is_permutation("a", "ab"));
        assert!(!is_permutation("a", "aaaaaa"));
        assert!(!is_permutation("ab", "abaaa"));
        assert!(!is_permutation("ab", "ab "));
        assert!(!is_permutation("abc", "abcd"));
    }
}
